src/programs/empire/expand.js

```diff
@@ -0,0 +1,365 @@
+
+    if (this.hostileSpawns.length > 0) {
+      return
+    }
+
+    // If the room isn't planned launch the room layout program, otherwise launch construction program
+
+    if (!this.colony.getLayout().isPlanned()) {
+      this.launchChildProcess('layout', 'city_layout', {
+        'room': this.data.colony
+      })
+    } else {
+      this.launchChildProcess('construct', 'city_construct', {
+        'room': this.data.colony
+      })
+    }
+
+    if (this.colony.getRoomSetting('SELF_SUFFICIENT')) {
+      if (!this.data.deathwatch) {
+        Room.addCity(this.data.colony)
+        this.data.deathwatch = Game.time
+      }
+      if (Game.time - this.data.deathwatch > 1800) {
+        this.suicide()
+      }
+    }
+  }
+
+  getNextCandidate () {
+    // If a candidate list already exists use the best scoring candidate from the list.
+    if (this.data.candidates && this.data.candidates.length > 0) {
+      return this.data.candidates.pop()
+    }
+
+    if (typeof this.data.candidateList === 'undefined') {
+      this.data.candidateList = this.getCandidateList()
+    }
+    if (!this.data.candidateScores) {
+      this.data.candidateScores = {}
+    }
+
+    const startCPU = Game.cpu.getUsed()
+    while (this.data.candidateList.length > 0) {
+      const testRoom = this.data.candidateList.pop()
+      const score = Room.getCityScore(testRoom)
+      if (score) {
+        this.data.candidateScores[testRoom] = score
+      }
+      if (Game.cpu.getUsed() - startCPU > 10) {
+        return
+      }
+    }
+
+    const scores = this.data.candidateScores
+    this.data.candidates = Object.keys(this.data.candidateScores)
+    this.data.candidates.sort(function (a, b) {
+      return scores[a] - scores[b]
+    })
+    return this.data.candidates.pop()
+  }
+
+  validateRoom (roomName) {
+    const closest = this.getClosestCity(roomName).name
+    const path = Game.map.findRoute(closest, roomName)
+    return path.length <= 8
+  }
+
+  getClosestCity (roomName) {
+    const cities = Room.getCities()
+    let closest = false
+    let distance = Infinity
+    for (let city of cities) {
+      if (!Game.rooms[city]) {
+        continue
+      }
+      const testDistance = Room.getManhattanDistance(city, roomName)
+      if (distance > testDistance) {
+        closest = city
+        distance = testDistance
+      }
+    }
+    return Game.rooms[closest]
+  }
+
+  getCandidateList () {
+    const cities = Room.getCities()
+    let candidates = []
+
+    for (let city of cities) {
+      if (!Game.rooms[city]) {
+        continue
+      }
+      const room = Game.rooms[city]
+      if (!room.getRoomSetting('EXPAND_FROM') || !room.isEconomyCapable('EXPAND_FROM')) {
+        continue
+      }
+
+      let rooms = Room.getRoomsInRange(city, 10)
+      for (let room of rooms) {
+        if (candidates.indexOf(room) >= 0 || !Room.isClaimable(room)) {
+          continue
+        }
+        if (Room.getManhattanDistance(city, room) <= 9) {
```

Comment:
This 9 seems to be somewhat at odds with the 8 in the validateRoom function.
Not a problem, but pointing it out to make sure it has been considered.

Reply:
I should probably change it in validateRoom, or make it a constant somewhere.